File: pebble/mobile-rn/modules/pebble-relay-crypto/ios/PebbleRelayCryptoModule.swift
import CryptoKit
import ExpoModulesCore
import Foundation
import Security

private let relayCryptoAlgorithm = "X25519-HKDF-SHA256-AES-256-GCM"
private let relayAssociatedData = "pebble.mobile-relay.v1"

public class PebbleRelayCryptoModule: Module {
  private var handshakes: [String: Curve25519.KeyAgreement.PrivateKey] = [:]
  private var sessions: [String: SymmetricKey] = [:]
  private let lock = NSLock()

  public func definition() -> ModuleDefinition {
    Name("PebbleRelayCrypto")

    AsyncFunction("createHandshake") { (input: [String: Any]) -> [String: Any] in
      let privateKey = Curve25519.KeyAgreement.PrivateKey()
      let handshakeId = UUID().uuidString
      let pairingSecretRef = input["pairingSecretRef"] as? String ?? ""
      let subscriptions = input["subscriptions"] as? [Any] ?? []
      let device = input["device"] as? [String: Any] ?? [:]

      lock.lock()
      handshakes[handshakeId] = privateKey
      lock.unlock()

      return [
        "handshakeId": handshakeId,
        "payload": [
          "device": device,
          "clientPublicKey": encodeRelayBase64(privateKey.publicKey.rawRepresentation),
          "pairingSecretRef": pairingSecretRef,
          "subscriptions": subscriptions
        ]
      ]
    }

    AsyncFunction("completeHandshake") { (input: [String: Any]) -> [String: Any] in
      guard let handshakeId = input["handshakeId"] as? String else {
        throw PebbleRelayCryptoError.invalidInput("handshakeId is required")
      }
      guard let ready = input["ready"] as? [String: Any] else {
        throw PebbleRelayCryptoError.invalidInput("ready is required")
      }
      guard ready["algorithm"] as? String == relayCryptoAlgorithm else {
        throw PebbleRelayCryptoError.invalidInput("unsupported relay crypto algorithm")
      }
      guard let expectedKeyId = ready["keyId"] as? String else {
        throw PebbleRelayCryptoError.invalidInput("ready.keyId is required")
      }
      guard let serverPublicKey = ready["serverPublicKey"] as? String else {
        throw PebbleRelayCryptoError.invalidInput("ready.serverPublicKey is required")
      }

      lock.lock()
      let privateKey = handshakes.removeValue(forKey: handshakeId)
      lock.unlock()

      guard let privateKey else {
        throw PebbleRelayCryptoError.invalidInput("unknown relay crypto handshake")
      }

      let publicKey = try Curve25519.KeyAgreement.PublicKey(
        rawRepresentation: try decodeRelayBase64(serverPublicKey)
      )
      let sharedSecret = try privateKey.sharedSecretFromKeyAgreement(with: publicKey)
      let key = deriveRelayKey(
        sharedSecret: sharedSecret,
        relayId: input["relayId"] as? String ?? "",
        pairingSecretRef: input["pairingSecretRef"] as? String ?? ""
      )
      let keyData = key.withUnsafeBytes { Data($0) }
      let keyId = relayKeyId(keyData)

      guard keyId == expectedKeyId else {
        throw PebbleRelayCryptoError.invalidInput("relay crypto key id mismatch")
      }

      let sessionId = UUID().uuidString
      lock.lock()
      sessions[sessionId] = key
      lock.unlock()

      return [
        "sessionId": sessionId,
        "keyId": keyId
      ]
    }

    AsyncFunction("encryptMessage") { (input: [String: Any]) -> [String: Any] in
      let key = try sessionKey(input["sessionId"] as? String)
      guard let message = input["message"] as? [String: Any] else {
        throw PebbleRelayCryptoError.invalidInput("message is required")
      }
      let plaintext = try JSONSerialization.data(withJSONObject: message, options: [])
      let nonceData = try randomBytes(count: 12)
      let nonce = try AES.GCM.Nonce(data: nonceData)
      let sealed = try AES.GCM.seal(
        plaintext,
        using: key,
        nonce: nonce,
        authenticating: Data(relayAssociatedData.utf8)
      )
      var ciphertext = Data(sealed.ciphertext)
      ciphertext.append(sealed.tag)

      return [
        "keyId": relayKeyId(key.withUnsafeBytes { Data($0) }),
        "nonce": encodeRelayBase64(nonceData),
        "ciphertext": encodeRelayBase64(ciphertext),
        "associatedData": relayAssociatedData
      ]
    }

    AsyncFunction("decryptMessage") { (input: [String: Any]) -> [String: Any] in
      let key = try sessionKey(input["sessionId"] as? String)
      guard let envelope = input["envelope"] as? [String: Any] else {
        throw PebbleRelayCryptoError.invalidInput("envelope is required")
      }
      guard envelope["keyId"] as? String == relayKeyId(key.withUnsafeBytes { Data($0) }) else {
        throw PebbleRelayCryptoError.invalidInput("relay crypto key id mismatch")
      }
      if let associatedData = envelope["associatedData"] as? String,
         associatedData != relayAssociatedData {
        throw PebbleRelayCryptoError.invalidInput("relay crypto associated data mismatch")
      }
      guard let nonceValue = envelope["nonce"] as? String else {
        throw PebbleRelayCryptoError.invalidInput("envelope.nonce is required")
      }
      guard let ciphertextValue = envelope["ciphertext"] as? String else {
        throw PebbleRelayCryptoError.invalidInput("envelope.ciphertext is required")
      }
      let combined = try decodeRelayBase64(ciphertextValue)
      guard combined.count >= 16 else {
        throw PebbleRelayCryptoError.invalidInput("encrypted relay payload is too short")
      }
      let ciphertext = combined.prefix(combined.count - 16)
      let tag = combined.suffix(16)
      let sealed = try AES.GCM.SealedBox(
        nonce: try AES.GCM.Nonce(data: try decodeRelayBase64(nonceValue)),
        ciphertext: ciphertext,
        tag: tag
      )
      let plaintext = try AES.GCM.open(
        sealed,
        using: key,
        authenticating: Data(relayAssociatedData.utf8)
      )
      guard let message = try JSONSerialization.jsonObject(with: plaintext, options: []) as? [String: Any] else {
        throw PebbleRelayCryptoError.invalidInput("encrypted relay payload was not an object")
      }

      return message
    }

    AsyncFunction("selfTest") { () -> [String: Any] in
      let relayId = "pebble-self-test-relay"
      let pairingSecretRef = "pebble-self-test-secret"
      let clientPrivateKey = Curve25519.KeyAgreement.PrivateKey()
      let serverPrivateKey = Curve25519.KeyAgreement.PrivateKey()
      let clientSharedSecret = try clientPrivateKey.sharedSecretFromKeyAgreement(
        with: serverPrivateKey.publicKey
      )
      let serverSharedSecret = try serverPrivateKey.sharedSecretFromKeyAgreement(
        with: clientPrivateKey.publicKey
      )
      let clientKey = deriveRelayKey(
        sharedSecret: clientSharedSecret,
        relayId: relayId,
        pairingSecretRef: pairingSecretRef
      )
      let serverKey = deriveRelayKey(
        sharedSecret: serverSharedSecret,
        relayId: relayId,
        pairingSecretRef: pairingSecretRef
      )
      let clientKeyId = relayKeyId(clientKey.withUnsafeBytes { Data($0) })
      let serverKeyId = relayKeyId(serverKey.withUnsafeBytes { Data($0) })

      guard clientKeyId == serverKeyId else {
        throw PebbleRelayCryptoError.invalidInput("relay crypto self-test key id mismatch")
      }

      let message: [String: Any] = [
        "version": relayAssociatedData,
        "id": "pebble-self-test-server",
        "type": "server.hello",
        "payload": [
          "relayId": relayId,
          "acceptedSubscriptions": [String]()
        ]
      ]
      let plaintext = try JSONSerialization.data(withJSONObject: message, options: [])
      let nonceData = try randomBytes(count: 12)
      let sealed = try AES.GCM.seal(
        plaintext,
        using: serverKey,
        nonce: try AES.GCM.Nonce(data: nonceData),
        authenticating: Data(relayAssociatedData.utf8)
      )
      var ciphertext = Data(sealed.ciphertext)
      ciphertext.append(sealed.tag)
      let opened = try AES.GCM.open(
        try AES.GCM.SealedBox(
          nonce: try AES.GCM.Nonce(data: nonceData),
          ciphertext: ciphertext.prefix(ciphertext.count - 16),
          tag: ciphertext.suffix(16)
        ),
        using: clientKey,
        authenticating: Data(relayAssociatedData.utf8)
      )
      guard let decoded = try JSONSerialization.jsonObject(with: opened, options: []) as? [String: Any],
            decoded["id"] as? String == "pebble-self-test-server" else {
        throw PebbleRelayCryptoError.invalidInput("relay crypto self-test decrypt mismatch")
      }

      return [
        "ok": true,
        "provider": "native",
        "algorithm": relayCryptoAlgorithm,
        "keyId": clientKeyId,
        "encryptedBytes": ciphertext.count
      ]
    }
  }

  private func sessionKey(_ sessionId: String?) throws -> SymmetricKey {
    guard let sessionId else {
      throw PebbleRelayCryptoError.invalidInput("sessionId is required")
    }
    lock.lock()
    let key = sessions[sessionId]
    lock.unlock()
    guard let key else {
      throw PebbleRelayCryptoError.invalidInput("unknown relay crypto session")
    }

    return key
  }
}

private enum PebbleRelayCryptoError: Error, LocalizedError {
  case invalidInput(String)

  var errorDescription: String? {
    switch self {
    case .invalidInput(let message):
      return message
    }
  }
}

private func deriveRelayKey(
  sharedSecret: SharedSecret,
  relayId: String,
  pairingSecretRef: String
) -> SymmetricKey {
  let salt = SHA256.hash(data: Data("pebble-mobile-relay:\(relayId):\(pairingSecretRef)".utf8))

  return sharedSecret.hkdfDerivedSymmetricKey(
    using: SHA256.self,
    salt: Data(salt),
    sharedInfo: Data(relayCryptoAlgorithm.utf8),
    outputByteCount: 32
  )
}

private func relayKeyId(_ keyData: Data) -> String {
  let digest = SHA256.hash(data: keyData)

  return encodeRelayBase64(Data(digest.prefix(16)))
}

private func randomBytes(count: Int) throws -> Data {
  var bytes = [UInt8](repeating: 0, count: count)
  let status = SecRandomCopyBytes(kSecRandomDefault, count, &bytes)
  guard status == errSecSuccess else {
    throw PebbleRelayCryptoError.invalidInput("secure random bytes unavailable")
  }

  return Data(bytes)
}

private func encodeRelayBase64(_ data: Data) -> String {
  data.base64EncodedString()
    .replacingOccurrences(of: "+", with: "-")
    .replacingOccurrences(of: "/", with: "_")
    .replacingOccurrences(of: "=", with: "")
}

private func decodeRelayBase64(_ value: String) throws -> Data {
  var normalized = value
    .replacingOccurrences(of: "-", with: "+")
    .replacingOccurrences(of: "_", with: "/")
  let padding = (4 - normalized.count % 4) % 4
  normalized.append(String(repeating: "=", count: padding))

  guard let data = Data(base64Encoded: normalized) else {
    throw PebbleRelayCryptoError.invalidInput("relay base64 value is invalid")
  }

  return data
}

File: pebble/mobile-rn/modules/pebble-relay-crypto/android/src/main/java/nebutra/pebble/relaycrypto/PebbleRelayCryptoModule.kt
package nebutra.pebble.relaycrypto

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.math.BigInteger
import java.security.KeyFactory
import java.security.KeyPairGenerator
import java.security.PrivateKey
import java.security.PublicKey
import java.security.SecureRandom
import java.security.interfaces.XECPublicKey
import java.security.spec.NamedParameterSpec
import java.security.spec.XECPublicKeySpec
import java.util.Base64
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap
import javax.crypto.Cipher
import javax.crypto.KeyAgreement
import javax.crypto.Mac
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec
import org.json.JSONArray
import org.json.JSONObject

private const val RELAY_CRYPTO_ALGORITHM = "X25519-HKDF-SHA256-AES-256-GCM"
private const val RELAY_ASSOCIATED_DATA = "pebble.mobile-relay.v1"

class PebbleRelayCryptoModule : Module() {
  private val handshakes = ConcurrentHashMap<String, PrivateKey>()
  private val sessions = ConcurrentHashMap<String, ByteArray>()
  private val secureRandom = SecureRandom()

  override fun definition() = ModuleDefinition {
    Name("PebbleRelayCrypto")

    AsyncFunction("createHandshake") { input: Map<String, Any?> ->
      val keyPairGenerator = KeyPairGenerator.getInstance("X25519")
      keyPairGenerator.initialize(NamedParameterSpec("X25519"))
      val keyPair = keyPairGenerator.generateKeyPair()
      val handshakeId = UUID.randomUUID().toString()
      handshakes[handshakeId] = keyPair.private

      mapOf(
        "handshakeId" to handshakeId,
        "payload" to mapOf(
          "device" to (input["device"] ?: emptyMap<String, Any?>()),
          "clientPublicKey" to encodeRelayBase64(x25519RawPublicKey(keyPair.public)),
          "pairingSecretRef" to (input["pairingSecretRef"] as? String ?: ""),
          "subscriptions" to (input["subscriptions"] ?: emptyList<Any>())
        )
      )
    }

    AsyncFunction("completeHandshake") { input: Map<String, Any?> ->
      val handshakeId = input["handshakeId"] as? String
        ?: throw IllegalArgumentException("handshakeId is required")
      val ready = input["ready"] as? Map<*, *>
        ?: throw IllegalArgumentException("ready is required")
      if (ready["algorithm"] as? String != RELAY_CRYPTO_ALGORITHM) {
        throw IllegalArgumentException("unsupported relay crypto algorithm")
      }
      val privateKey = handshakes.remove(handshakeId)
        ?: throw IllegalArgumentException("unknown relay crypto handshake")
      val serverPublicKey = ready["serverPublicKey"] as? String
        ?: throw IllegalArgumentException("ready.serverPublicKey is required")
      val expectedKeyId = ready["keyId"] as? String
        ?: throw IllegalArgumentException("ready.keyId is required")
      val keyAgreement = KeyAgreement.getInstance("X25519")
      keyAgreement.init(privateKey)
      keyAgreement.doPhase(x25519PublicKeyFromRaw(decodeRelayBase64(serverPublicKey)), true)
      val keyBytes = deriveRelayKey(
        sharedSecret = keyAgreement.generateSecret(),
        relayId = input["relayId"] as? String ?: "",
        pairingSecretRef = input["pairingSecretRef"] as? String ?: ""
      )
      val keyId = relayKeyId(keyBytes)
      if (keyId != expectedKeyId) {
        throw IllegalArgumentException("relay crypto key id mismatch")
      }
      val sessionId = UUID.randomUUID().toString()
      sessions[sessionId] = keyBytes

      mapOf(
        "sessionId" to sessionId,
        "keyId" to keyId
      )
    }

    AsyncFunction("encryptMessage") { input: Map<String, Any?> ->
      val sessionId = input["sessionId"] as? String
        ?: throw IllegalArgumentException("sessionId is required")
      val keyBytes = sessions[sessionId]
        ?: throw IllegalArgumentException("unknown relay crypto session")
      val message = input["message"] as? Map<*, *>
        ?: throw IllegalArgumentException("message is required")
      val nonce = ByteArray(12)
      secureRandom.nextBytes(nonce)
      val cipher = Cipher.getInstance("AES/GCM/NoPadding")
      cipher.init(Cipher.ENCRYPT_MODE, SecretKeySpec(keyBytes, "AES"), GCMParameterSpec(128, nonce))
      cipher.updateAAD(RELAY_ASSOCIATED_DATA.toByteArray(Charsets.UTF_8))
      val ciphertext = cipher.doFinal(JSONObject(message).toString().toByteArray(Charsets.UTF_8))

      mapOf(
        "keyId" to relayKeyId(keyBytes),
        "nonce" to encodeRelayBase64(nonce),
        "ciphertext" to encodeRelayBase64(ciphertext),
        "associatedData" to RELAY_ASSOCIATED_DATA
      )
    }

    AsyncFunction("decryptMessage") { input: Map<String, Any?> ->
      val sessionId = input["sessionId"] as? String
        ?: throw IllegalArgumentException("sessionId is required")
      val keyBytes = sessions[sessionId]
        ?: throw IllegalArgumentException("unknown relay crypto session")
      val envelope = input["envelope"] as? Map<*, *>
        ?: throw IllegalArgumentException("envelope is required")
      if (envelope["keyId"] as? String != relayKeyId(keyBytes)) {
        throw IllegalArgumentException("relay crypto key id mismatch")
      }
      val associatedData = envelope["associatedData"] as? String
      if (associatedData != null && associatedData != RELAY_ASSOCIATED_DATA) {
        throw IllegalArgumentException("relay crypto associated data mismatch")
      }
      val nonce = decodeRelayBase64(
        envelope["nonce"] as? String ?: throw IllegalArgumentException("envelope.nonce is required")
      )
      val ciphertext = decodeRelayBase64(
        envelope["ciphertext"] as? String ?: throw IllegalArgumentException("envelope.ciphertext is required")
      )
      val cipher = Cipher.getInstance("AES/GCM/NoPadding")
      cipher.init(Cipher.DECRYPT_MODE, SecretKeySpec(keyBytes, "AES"), GCMParameterSpec(128, nonce))
      cipher.updateAAD(RELAY_ASSOCIATED_DATA.toByteArray(Charsets.UTF_8))
      val plaintext = String(cipher.doFinal(ciphertext), Charsets.UTF_8)

      JSONObject(plaintext).toMap()
    }

    AsyncFunction("selfTest") {
      selfTestRelayCrypto()
    }
  }

  private fun selfTestRelayCrypto(): Map<String, Any> {
    val relayId = "pebble-self-test-relay"
    val pairingSecretRef = "pebble-self-test-secret"
    val keyPairGenerator = KeyPairGenerator.getInstance("X25519")
    keyPairGenerator.initialize(NamedParameterSpec("X25519"))
    val clientKeyPair = keyPairGenerator.generateKeyPair()
    val serverKeyPair = keyPairGenerator.generateKeyPair()
    val clientKeyBytes = deriveRelayKey(
      sharedSecret = sharedSecret(clientKeyPair.private, serverKeyPair.public),
      relayId = relayId,
      pairingSecretRef = pairingSecretRef
    )
    val serverKeyBytes = deriveRelayKey(
      sharedSecret = sharedSecret(serverKeyPair.private, clientKeyPair.public),
      relayId = relayId,
      pairingSecretRef = pairingSecretRef
    )
    val keyId = relayKeyId(clientKeyBytes)

    if (keyId != relayKeyId(serverKeyBytes)) {
      throw IllegalArgumentException("relay crypto self-test key id mismatch")
    }

    val message = mapOf(
      "version" to RELAY_ASSOCIATED_DATA,
      "id" to "pebble-self-test-server",
      "type" to "server.hello",
      "payload" to mapOf(
        "relayId" to relayId,
        "acceptedSubscriptions" to emptyList<String>()
      )
    )
    val nonce = ByteArray(12)
    secureRandom.nextBytes(nonce)
    val cipher = Cipher.getInstance("AES/GCM/NoPadding")
    cipher.init(Cipher.ENCRYPT_MODE, SecretKeySpec(serverKeyBytes, "AES"), GCMParameterSpec(128, nonce))
    cipher.updateAAD(RELAY_ASSOCIATED_DATA.toByteArray(Charsets.UTF_8))
    val ciphertext = cipher.doFinal(JSONObject(message).toString().toByteArray(Charsets.UTF_8))
    val decryptCipher = Cipher.getInstance("AES/GCM/NoPadding")
    decryptCipher.init(Cipher.DECRYPT_MODE, SecretKeySpec(clientKeyBytes, "AES"), GCMParameterSpec(128, nonce))
    decryptCipher.updateAAD(RELAY_ASSOCIATED_DATA.toByteArray(Charsets.UTF_8))
    val plaintext = String(decryptCipher.doFinal(ciphertext), Charsets.UTF_8)
    val decoded = JSONObject(plaintext)

    if (decoded.getString("id") != "pebble-self-test-server") {
      throw IllegalArgumentException("relay crypto self-test decrypt mismatch")
    }

    return mapOf(
      "ok" to true,
      "provider" to "native",
      "algorithm" to RELAY_CRYPTO_ALGORITHM,
      "keyId" to keyId,
      "encryptedBytes" to ciphertext.size
    )
  }

  private fun deriveRelayKey(
    sharedSecret: ByteArray,
    relayId: String,
    pairingSecretRef: String
  ): ByteArray {
    val saltInput = "pebble-mobile-relay:$relayId:$pairingSecretRef".toByteArray(Charsets.UTF_8)
    val salt = java.security.MessageDigest.getInstance("SHA-256").digest(saltInput)
    val prk = hmacSha256(salt, sharedSecret)
    val info = RELAY_CRYPTO_ALGORITHM.toByteArray(Charsets.UTF_8)

    return hmacSha256(prk, info + byteArrayOf(1)).copyOfRange(0, 32)
  }

  private fun hmacSha256(key: ByteArray, input: ByteArray): ByteArray {
    val mac = Mac.getInstance("HmacSHA256")
    mac.init(SecretKeySpec(key, "HmacSHA256"))

    return mac.doFinal(input)
  }

  private fun relayKeyId(keyBytes: ByteArray): String {
    val digest = java.security.MessageDigest.getInstance("SHA-256").digest(keyBytes)

    return encodeRelayBase64(digest.copyOfRange(0, 16))
  }

  private fun sharedSecret(privateKey: PrivateKey, publicKey: PublicKey): ByteArray {
    val keyAgreement = KeyAgreement.getInstance("X25519")
    keyAgreement.init(privateKey)
    keyAgreement.doPhase(publicKey, true)

    return keyAgreement.generateSecret()
  }
}

private fun x25519RawPublicKey(publicKey: PublicKey): ByteArray {
  val u = (publicKey as XECPublicKey).u
  val bigEndian = u.toByteArray()
  val raw = ByteArray(32)

  for (index in bigEndian.indices) {
    val target = index
    val source = bigEndian.size - 1 - index
    if (target < raw.size) {
      raw[target] = bigEndian[source]
    }
  }

  return raw
}

private fun x25519PublicKeyFromRaw(raw: ByteArray): PublicKey {
  val u = BigInteger(1, raw.reversedArray())
  val spec = XECPublicKeySpec(NamedParameterSpec("X25519"), u)

  return KeyFactory.getInstance("X25519").generatePublic(spec)
}

private fun encodeRelayBase64(value: ByteArray): String =
  Base64.getUrlEncoder().withoutPadding().encodeToString(value)

private fun decodeRelayBase64(value: String): ByteArray =
  Base64.getUrlDecoder().decode(value)

private fun JSONObject.toMap(): Map<String, Any?> {
  val output = mutableMapOf<String, Any?>()
  val keys = keys()

  while (keys.hasNext()) {
    val key = keys.next()
    output[key] = unwrapJsonValue(get(key))
  }

  return output
}

private fun JSONArray.toListValue(): List<Any?> {
  val output = mutableListOf<Any?>()

  for (index in 0 until length()) {
    output.add(unwrapJsonValue(get(index)))
  }

  return output
}

private fun unwrapJsonValue(value: Any?): Any? =
  when (value) {
    JSONObject.NULL -> null
    is JSONObject -> value.toMap()
    is JSONArray -> value.toListValue()
    else -> value
  }
